packages/react-combobox/src/components/ComboButton/useComboButton.ts

```diff
@@ -11,18 +14,50 @@ import type { ComboButtonProps, ComboButtonState } from './ComboButton.types';
  * @param props - props from this instance of ComboButton
  * @param ref - reference to root HTMLElement of ComboButton
  */
-export const useComboButton_unstable = (props: ComboButtonProps, ref: React.Ref<HTMLElement>): ComboButtonState => {
+export const useComboButton_unstable = (
+  props: ComboButtonProps,
+  ref: React.Ref<HTMLButtonElement>,
+): ComboButtonState => {
+  const { placeholder, value } = props;
+
+  const { activeOption, open } = useContextSelector(ComboboxContext, ctx => ({
+    activeOption: ctx.activeOption,
+    open: ctx.open,
+  }));
```

Comment:
The purpose of `useContextSelector()` is to skip some update, i.e. update only when a `value` (_result of selector function_) will change.

With current implementation we will re-render this component every time when `activeOption` changes. Implementation-wise it's correct as we depend on it for `aria-activedescendant`, but performance-wise it's not the best approach.

Reply:
with aria-activedescendant, I think we do not have any other option, we have to re-render components that change the active state. We have to use activedescendant, because focus needs to stay in the Input to allow user to type. Do you see any other option?

Reply:
Returning a new object here will result in more rerenders than if you called `useContextSelector` on primitive values because the object reference is not stable and a new object is created on every update.

this one is a nit: I would also suggest to wrap the usage of `useContextSelector` to `useComboboxContext` or smth similar to make future typing easier :)

Reply:
> with aria-activedescendant, I think we do not have any other option, we have to re-render components that change the active state. We have to use activedescendant, because focus needs to stay in the Input to allow user to type. Do you see any other option?

It should be possible IMO hook into native DOM at this point to make these updates and avoid the React render lifecycle as a perf optimization. Would need to prototype to see how it would work... but I'm quite sure it would be possible

Reply:
> It should be possible IMO hook into native DOM at this point to make these updates and avoid the React render lifecycle as a perf optimization. Would need to prototype to see how it would work and properly evaluate the pros/cons... but I'm quite sure it would be possible

Yes, it's something that I proposed initially. It's a bit against React's nature, but allows to avoid updates.

Reply:
I've updated an original prototype made by @layershifter  that includes the `aria-activedescendant` feature and reduced the renders for option components.

https://codesandbox.io/s/staging-glitter-vmgsky?file=/src/App.tsx

The result is that, during keyboard navigation only maximum 2 combobox options will rerender and the combobox itself does not rerender

Reply:
I've opened an issue for this so it can be addressed outside this PR: #22055, and added it to the component implementation checklist so it won't be missed: https://github.com/microsoft/fluentui/issues/18133